### PR TITLE
[Feature][Branch][Snackbar] Consecutive snackbars

### PR DESCRIPTION
**Pull request by JongtaekChoi**
## Description

I modified the way of showing `snackbar`.
A snack bar has a `show` function and will be appeared when the function is called. 
If the function called when the previous one is showing, then the previous one will be dismissed immediately and the next one will appear after that. 

Jan-08-2020 00-36-57

## Related Issues

#46 

## Tests

I added the following tests:

- simulating show function called. 

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/dooboo-ui-native/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

**Comment by Ryz0nd** [comment]
LGTM!!

**Comment by taejs** [comment]
From the user's point of view, there might be some possibility that user can't check the previous Snackbar. Because it disappear as soon as another Snackbar showing up.

Referring to [behavior](https://material.io/components/snackbars/#behavior), in Appearing and disappearing section
>  They automatically disappear from the screen after a minimum of four seconds, and a maximum of ten seconds.
So Snackbar need disappear with some delay

Also, this issue can be related to the Principles section of behavior.
> Snackbars appear temporarily, and disappear on their own without requiring user input to be dismissed.

Consider implementing this principle when you're working on it.

**Comment by JongtaekChoi** [comment]
@Ta2Rim 
Thanks for your opinion. 
But, I think the previous Snackbar should disappear as soon as another Snackbar showing up.
 
The document says, 
> Only one Snackbar may be displayed at a time.

If a previous Snackbar exists, a new one can appear only after the previous one disappeared.

But it is weird that Snackbar like below appears some seconds after user action occurred. 
image
undoimage

So, I think the phrase, `(A Snackbar) disappear on their own without requiring user input to be dismissed.` does not mean it **always** should show until it disappeared itself. 

Let's see this video.
[image](https://storage.googleapis.com/spec-host-backup/mio-components%2Fassets%2F1smqQHvwlE8Nm5nNwHXgvnJ9JDrOt_CB1%2F01-appear.mp4)
It does not need a 'dismiss' button but when the user clicks the 'RETRY' button it disappeared.

A Snackbar is a temporary UI component and has low priority. So, I think that if the user started another action, it would be OK to disappear immediately.

If you have another opinion, please let me know again.

**Comment by taejs** [comment]
You're right, it's right behavior that showing immediately the result of user action. Thanks for explanation.

**Comment by taejs** [comment]
There's another question. It looks Snackbar component always be single, and the parent component just changes its properties toggling `show` variable. right?
I think it's brilliant! For now, there's no need to manage a bunch of Snackbars because the Snackbar with last property is only shown.
I just wonder why you implement this feature like that.

**Comment by JongtaekChoi** [comment]
@Ta2Rim 
I just thought it was better a Snackbar provide show function than the parent View manages the show state of the Snackbar.

**Comment by godon019** [comment]
using imperative handle to show snackbar like that sounds awesome for its behavior!

File: src/components/shared/__tests__/Snackbar.test.tsx [hunk 1]
```diff
@@ -1,41 +1,47 @@
...
-  it('should dismiss after timeout', async () => {
-    render(component);
+  it('should simulate showing snackbar', async () => {
+    const renderResult = render(<TestWrapper />);
+    const btn = renderResult.getByTestId('Button');
+    act(() => {
+      fireEvent.press(btn);
+      fireEvent.press(btn);
+    });
     jest.runAllTimers();
```
Comment (godon019) [inline src/components/shared/__tests__/Snackbar.test.tsx:45]: It seems like the test is not completely written, maybe you missed it?

Comment (godon019) [reply src/components/shared/__tests__/Snackbar.test.tsx:45]: I am not sure about the ref's behavior but I guess there might be some way to check if the `show()` function in ref.current got called. how do you think so?

File: src/components/shared/Snackbar/index.tsx [hunk 2]
```diff
@@ -38,52 +41,95 @@ const ActionButton = styled.View`
...
   React.useEffect(() => {
-    let timeout;
-    if (show === true) {
-      timeout = setTimeout(() => {
-        setShow(false);
-      }, timer);
+    if (isShowing) {
+      if (isVisible) {
+        close();
+      } else {
```
Comment (godon019) [inline src/components/shared/Snackbar/index.tsx:98]: I don't see in which situation that case will be met, could you elaborate a bit for me please?

Comment (godon019) [reply src/components/shared/Snackbar/index.tsx:98]: seems like there is no `isShowing === true && isVisible ===true` case in the whole process

Comment (JongtaekChoi) [reply src/components/shared/Snackbar/index.tsx:98]: You are right. It was my mistake. I'll fix it.

File: src/components/shared/__tests__/Snackbar.test.tsx (hunk 1 above)
Comment (JongtaekChoi) [reply src/components/shared/__tests__/Snackbar.test.tsx:45]: I'll add checking codes. Thanks!

**Pull request merged**